### PR TITLE
[PyTorch] PyTorch 1.10.2 SageMaker DLC

### PR DESCRIPTION
**Pull request by waytrue17**
*GitHub Issue #, if available:*

Note: If merging this PR should also close the associated Issue, please also add that Issue # to the Linked Issues section on the right. 

### Tests run
**NOTE: If you are creating a PR for a new framework version, please ensure success of the standard, rc, and efa sagemaker remote tests by updating the dlc_developer_config.toml file:**

- [x] Revision A: `sagemaker_remote_tests = "standard"`
- [x] Revision B: `sagemaker_remote_tests = "rc"`
- [x] Revision C: `sagemaker_remote_tests = "efa"`

**Additionally, please run the sagemaker local tests in at least one revision:**
- [x] `sagemaker_local_tests = true`

## Label Checklist
- [ ] I have added the project label for this PR (*<project_name>* or "Improvement")

## PR Checklist
- [ ] I've prepended PR tag with frameworks/job this applies to : [mxnet, tensorflow, pytorch] | [ei/neuron/graviton] | [build] | [test] | [benchmark] | [ec2, ecs, eks, sagemaker]
- [ ] If the PR changes affects SM test, I've modified dlc_developer_config.toml in my PR branch by setting sagemaker_tests = true and efa_tests = true
- [ ] If this PR changes existing code, the change fully backward compatible with pre-existing code. (Non backward-compatible changes need special approval.)
- [ ] (If applicable) I've documented below the DLC image/dockerfile this relates to
- [ ] (If applicable) I've documented below the tests I've run on the DLC image
- [ ] (If applicable) I've reviewed the licenses of updated and new binaries and their dependencies to make sure all licenses are on the Apache Software Foundation Third Party License Policy Category A or Category B license list.  See [https://www.apache.org/legal/resolved.html](https://www.apache.org/legal/resolved.html).
- [ ] (If applicable) I've scanned the updated and new binaries to make sure they do not have vulnerabilities associated with them.

## Pytest Marker Checklist
- [ ] (If applicable) I have added the marker `@pytest.mark.model("<model-type>")` to the new tests which I have added, to specify the Deep Learning model that is used in the test (use `"N/A"` if the test doesn't use a model)
- [ ] (If applicable) I have added the marker `@pytest.mark.integration("<feature-being-tested>")` to the new tests which I have added, to specify the feature that will be tested
- [ ] (If applicable) I have added the marker `@pytest.mark.multinode(<integer-num-nodes>)` to the new tests which I have added, to specify the number of nodes used on a multi-node test
- [ ] (If applicable) I have added the marker `@pytest.mark.processor(<"cpu"/"gpu"/"eia"/"neuron">)` to the new tests which I have added, if a test is specifically applicable to only one processor type

#### EIA/NEURON/GRAVITON Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `ei_mode = true`, `neuron_mode = true` or `graviton_mode = true`

#### Benchmark Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `benchmark_mode = true`

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license. I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: test/sagemaker_tests/pytorch/training/resources/mnist/smdataparallel_mnist.py [hunk 1]
```diff
@@ -135,7 +137,7 @@ def main():
     args = parser.parse_args()
     args.world_size = dist.get_world_size()
     args.rank = rank = dist.get_rank()
-    args.local_rank = local_rank = dist.get_local_rank()
+    args.local_rank = local_rank = os.environ['OMPI_COMM_WORLD_LOCAL_RANK']
```
Comment (roywei) [inline test/sagemaker_tests/pytorch/training/resources/mnist/smdataparallel_mnist.py:140]:
```suggestion
    args.local_rank = local_rank = int(os.environ['OMPI_COMM_WORLD_LOCAL_RANK'])
```

**Review by roywei — COMMENTED** [review]
Sorry the SMDDP mnist test need to updated, you commit the change suggested. Thanks!

File: test/sagemaker_tests/pytorch/training/resources/mnist/smdataparallel_mnist.py (hunk 1 above)
Comment (waytrue17) [reply test/sagemaker_tests/pytorch/training/resources/mnist/smdataparallel_mnist.py:140]: Thanks for the update

File: test/sagemaker_tests/pytorch/training/resources/mnist/smdataparallel_smmodelparallel_mnist_script_mode.sh [hunk 2]
```diff
@@ -4,6 +4,6 @@
 
 set -ex
 
-smddpsinglenode python smdataparallel_mnist.py
+smddp_singlenode_dev python smdataparallel_mnist.py
```
Comment (roywei) [inline test/sagemaker_tests/pytorch/training/resources/mnist/smdataparallel_smmodelparallel_mnist_script_mode.sh:7]:
```suggestion
smddprun python smdataparallel_mnist.py
```

File: test/sagemaker_tests/tensorflow/tensorflow2_training/resources/smmodelparallel/smdataparallel_smmodelparallel_mnist_script_mode.sh [hunk 3]
```diff
@@ -4,5 +4,5 @@
 
 set -ex
 
-smddpsinglenode python smdataparallel_mnist.py
+smddp_singlenode_dev python smdataparallel_mnist.py
```
Comment (roywei) [inline test/sagemaker_tests/tensorflow/tensorflow2_training/resources/smmodelparallel/smdataparallel_smmodelparallel_mnist_script_mode.sh:7]:
```suggestion
smddprun python smdataparallel_mnist.py
```

File: test/sagemaker_tests/tensorflow/tensorflow2_training/resources/smmodelparallel/smdataparallel_smmodelparallel_mnist_script_mode.sh [hunk 4]
```diff
@@ -4,5 +4,6 @@
...
```
Comment (shantanutrip) [inline test/sagemaker_tests/tensorflow/tensorflow2_training/resources/smmodelparallel/smdataparallel_smmodelparallel_mnist_script_mode.sh:7]: I am not sure what the first argument does and why have we changed it. Can you please explain?

File: pytorch/training/docker/1.10/py3/cu113/Dockerfile.sagemaker.gpu [hunk 5]
```diff
@@ -51,12 +50,19 @@ COPY start_with_right_hostname.sh /usr/local/bin/start_with_right_hostname.sh
...
-    smdebug==${SMDEBUG_VERSION} \
+    # smdebug==${SMDEBUG_VERSION} \
     smclarify \
     "sagemaker>=2,<3" \
     sagemaker-experiments==0.* \
     sagemaker-pytorch-training
 
+# Install smdebug from souce for test purpose
+RUN cd /tmp \
+  && git clone -b 1.0.13 https://github.com/awslabs/sagemaker-debugger \
```
Comment (shantanutrip) [inline pytorch/training/docker/1.10/py3/cu113/Dockerfile.sagemaker.gpu:61]: I am seeing that the `SMDEBUG_VERSION` used above is 1.0.12, however we are checking out to 1.0.13, is that on purpose?

File: pytorch/training/docker/1.10/py3/Dockerfile.sagemaker.cpu [hunk 6]
```diff
@@ -32,14 +32,21 @@ RUN conda install -y \
...
     smclarify \
     "sagemaker>=2,<3" \
     sagemaker-experiments==0.* \
     "sagemaker-pytorch-training<3" \
  && pip uninstall -y torch \
  && pip install --no-cache-dir -U ${PT_TRAINING_URL}
 
+# Install smdebug from souce for test purpose
+RUN cd /tmp \
+  && git clone -b 1.0.13 https://github.com/awslabs/sagemaker-debugger \
```
Comment (shantanutrip) [inline pytorch/training/docker/1.10/py3/Dockerfile.sagemaker.cpu:45]: I am seeing that the SMDEBUG_VERSION used above is 1.0.12, however we are checking out to 1.0.13, is that on purpose?

File: test/sagemaker_tests/tensorflow/tensorflow2_training/resources/smmodelparallel/smdataparallel_smmodelparallel_mnist_script_mode.sh (hunk 4 above)
Comment (waytrue17) [reply test/sagemaker_tests/tensorflow/tensorflow2_training/resources/smmodelparallel/smdataparallel_smmodelparallel_mnist_script_mode.sh:7]: SMDDP team suggested to make such change. My understanding is smddp launcher of single node is only for development purpose. @roywei may have more context here.

File: pytorch/training/docker/1.10/py3/cu113/Dockerfile.sagemaker.gpu (hunk 5 above)
Comment (waytrue17) [reply pytorch/training/docker/1.10/py3/cu113/Dockerfile.sagemaker.gpu:61]: SMDEBUG team suggested to make such change. smdebug 1.0.12 does not support PT1.10.2, upgrading to 1.0.13 for the support

File: pytorch/training/docker/1.10/py3/Dockerfile.sagemaker.cpu (hunk 6 above)
Comment (waytrue17) [reply pytorch/training/docker/1.10/py3/Dockerfile.sagemaker.cpu:45]: SMDEBUG team suggested to make such change. smdebug 1.0.12 does not support PT1.10.2, upgrading to 1.0.13 for the support

Comment (shantanutrip) [reply pytorch/training/docker/1.10/py3/Dockerfile.sagemaker.cpu:45]: I think it is a bit confusing to see `SMDEBUG_VERSION=1.0.12` and then cloning branch `1.0.13`. `SMDEBUG_VERSION` is not being used anywhere in the dockerfile. Why don't we change it to `1.0.13` and use it in the git clone command?

Comment (waytrue17) [reply pytorch/training/docker/1.10/py3/Dockerfile.sagemaker.cpu:45]: I think smdebug team has not release 1.0.13 into PyPi. So the only way is to install from source code. This is also the way smdebug team suggested.

Comment (shantanutrip) [reply pytorch/training/docker/1.10/py3/Dockerfile.sagemaker.cpu:45]: Yes, I am agreeing to conduct the installation from the source code. I am just concerned that if we leave the `SMDEBUG_VERSION=1.0.12`, in the future, we might get confused by looking at this variable. So, we could make the following changes:
1. `SMDEBUG_VERSION=1.0.13`
2. git clone -b ${SMDEBUG_VERSION} https://github.com/awslabs/sagemaker-debugger

Comment (waytrue17) [reply pytorch/training/docker/1.10/py3/Dockerfile.sagemaker.cpu]: Done

File: test/sagemaker_tests/tensorflow/tensorflow2_training/resources/smmodelparallel/smdataparallel_smmodelparallel_mnist_script_mode.sh [hunk 7]
```diff
@@ -4,5 +4,6 @@
```
Comment (shantanutrip) [inline test/sagemaker_tests/tensorflow/tensorflow2_training/resources/smmodelparallel/smdataparallel_smmodelparallel_mnist_script_mode.sh:4]: This is a Tensorflow file, why have we made change to this?

Comment (waytrue17) [reply test/sagemaker_tests/tensorflow/tensorflow2_training/resources/smmodelparallel/smdataparallel_smmodelparallel_mnist_script_mode.sh:4]: This change is also suggested by SMDDP team. The reason is to update the single node launcher that being used in both TF and PT tests.

Comment (shantanutrip) [reply test/sagemaker_tests/tensorflow/tensorflow2_training/resources/smmodelparallel/smdataparallel_smmodelparallel_mnist_script_mode.sh:4]: Thanks a lot!

**Review by shantanutrip — DISMISSED** [review]
Looks good to me. Let us handle the SMProfiler cases on the mainline as we had discussed earlier. Once the builds complete, please merge the PR.

File: pytorch/buildspec.yml [hunk 8]
```diff
@@ -71,34 +71,34 @@ images:
...
-  #   python_version: &DOCKER_PYTHON_VERSION py3
-  #   tag_python_version: &TAG_PYTHON_VERSION py38
-  #   cuda_version: &CUDA_VERSION cu113
-  #   os_version: &OS_VERSION ubuntu20.04
-  #   tag: !join [ *VERSION, "-", *DEVICE_TYPE, "-", *TAG_PYTHON_VERSION, "-", *CUDA_VERSION, "-", *OS_VERSION, "-sagemaker" ]
-  #   docker_file: !join [ docker/, *SHORT_VERSION, /, *DOCKER_PYTHON_VERSION, /, *CUDA_VERSION, /Dockerfile.sagemaker.,
-  #                        *DEVICE_TYPE ]
-  #   context:
-  #     <<: *TRAINING_CONTEXT
+  BuildSageMakerCPUPTTrainPy3DockerImage:
```
[the remark this answers is not in the document]
Comment (waytrue17) [reply pytorch/buildspec.yml:76]: Done

**Review by shantanutrip — APPROVED** [review]
LGTM!!  Let us handle the SMProfiler cases on the mainline as we had discussed earlier. Once the builds complete, please merge the PR.

**Review by ydaiming — APPROVED** [review]

**Review by roywei — APPROVED** [review]
LGTM from SMDDP side, thanks!

**Review by josephevans — APPROVED** [review]

**Comment by shantanutrip** [comment]
Merging the PR in interest of time. However, the review from SMDebug is still required before the release.

**Review by MZSHAN — COMMENTED** [review]
Looks good to me.

**Pull request merged**